source/server/listener_manager_impl.cc

```diff
@@ -109,25 +109,38 @@ ProdListenerComponentFactory::createDrainManager(envoy::api::v2::Listener::Drain
 // This same object can be extended to handle additional listener socket options.
 class ListenerSocketOption : public Network::Socket::Option, Logger::Loggable<Logger::Id::config> {
 public:
-  ListenerSocketOption(bool transparent) : transparent_(transparent) {}
+  ListenerSocketOption(const envoy::api::v2::Listener& config)
+      : transparent_(config.transparent()) {}
 
   // Network::Socket::Option
-  bool setOption(Network::Socket& socket, bool pre_bind) const override {
-    // IP_TRANSPARENT makes a difference only for the following bind(). Set it before the bind
-    // and ignore it after.
-    if (transparent_ && pre_bind) {
+  bool setOption(Network::Socket& socket, Network::Socket::SocketState state) const override {
+    // IP_TRANSPARENT has an effect on bind() (allowing bind() to non-local addresses), but
+    // also after bind(), controlling whether new connections with non-local addresses are
+    // redirected to the listening socket or not, so we need to set it in either case, as
+    // it is possible to, e.g., set it on a listener and then change the listener configuration
+    // to not set it.
+    // This has the side effect of setting the option twice after the socket is first created,
+    // both before and after bind().
+    UNREFERENCED_PARAMETER(state);
+    int transparent = transparent_;
+
 #ifdef SOL_IP
-      int on = 1;
-      if (setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &on, sizeof(on)) == -1) {
-        ENVOY_LOG(warn, "Setting IP_TRANSPARENT on listener socket failed: {}", strerror(errno));
-        throw EnvoyException("ListenSocketOption: Error setting IP_TRANSPARENT socket option");
-      }
+    int rc = setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &transparent, sizeof(transparent));
+    // If we are unable to set the option we still return success for the default value (0),
+    // as we never changed the option to begin with. This also prevents failures in all current
+    // test cases that do not set the transparent option.
+    if (rc == 0 || transparent == 0) {
```

Comment:
Is there a case where rc != 0, transparent == 0, and transparent_ != 0? In other words, will setsockopt ever modify transparent?  If not, I think it's clearer if you check against transparent_ here, so it's clear that the result doesn't depend on setsockopt possibly modifying it.

Reply:
You are right, will compare against `transparent_`